### PR TITLE
Added an initial implementation of partitioned communication over persistant MPI calls.

### PR DESCRIPTION
**Pull request by mdosanjh**
Signed-off-by: Matthew Dosanjh <mdosanj@sandia.gov>

This is an initial implementation of partitioned communication using persistent sends and persistent receives. This should be compliant with the current release candidate of the MPI-4.0 specification.

### CODE REVIEW

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/03dace16a79e5958d962c52e752aee8d

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/6e3e4e117f7c2aefd199166acefc6be9

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/f6049cb406966176744f98a0685e5655

**Comment by mdosanjh** [comment]
bot:retest

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/3fc96a30c7741cd755a49eafcc7be691

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/eb81ff3fe01960c95f32f78d509140d3

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/54c78e4f404b5a0637d79fc9793b2ae0

**Comment by mdosanjh** [comment]
bot:retest

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/7a7ba36c4d7585b1384612d9500095b2

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/fec21300f6fdbde084d421c53690c9c3

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/fbe93757a0df3d339c7e34f226c21ebd

**Comment by mdosanjh** [comment]
I'm having trouble parsing these errors from the checks (one seems to be happening in OB1?) does anyone have insight on to what I need to fix?

**Comment by mdosanjh** [comment]
bot:retest

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/08068d909e505586843f9b9497d51a1e

**Comment by jsquyres** [comment]
@mdosanjh Not sure why you're asking about ob1, the error is later:

```
Making all in mpi/fortran/use-mpi-f08/bindings
make[2]: Entering directory '/tmp/ompi-src/ompi/mpi/fortran/use-mpi-f08/bindings'
  PPFC     ompi-mpifh-bindings.lo
"mpi-f-interfaces-bind.h", line 331.28: 1513-230 (S) Procedure ompi_parrived_f was declared with the BIND(C) attribute.  Dummy argument flag of ompi_parrived_f must have interoperable type and type parameters.
** ompi_mpifh_bindings   === End of Compilation 1 ===
1501-511  Compilation failed for file ompi-mpifh-bindings.F90.
```

**Comment by mdosanjh** [comment]
@jsquyres Apologies, this is my first time interacting with the pull request check system and ob1 was where https://gist.github.com/08068d909e505586843f9b9497d51a1e stopped/truncated as far as I could tell, so I had assumed that was where the issue was. I have subsequently found a local system to replicate the error on and I believe I have fixated it

> @mdosanjh Not sure why you're asking about ob1, the error is later:
> 
> ```
> Making all in mpi/fortran/use-mpi-f08/bindings
> make[2]: Entering directory '/tmp/ompi-src/ompi/mpi/fortran/use-mpi-f08/bindings'
>   PPFC     ompi-mpifh-bindings.lo
> "mpi-f-interfaces-bind.h", line 331.28: 1513-230 (S) Procedure ompi_parrived_f was declared with the BIND(C) attribute.  Dummy argument flag of ompi_parrived_f must have interoperable type and type parameters.
> ** ompi_mpifh_bindings   === End of Compilation 1 ===
> 1501-511  Compilation failed for file ompi-mpifh-bindings.F90.
> ```

File: ompi/mca/part/base/part_base_frame.c [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+#define xstringify(part) #part
+#define stringify(part) xstringify(part)
+
+/*
+ * Global variables
+ */
+mca_part_base_module_t mca_part = {
+    mca_part_base_progress,   /* part_progress */
```
Comment (hjelmn) [inline ompi/mca/part/base/part_base_frame.c:70]: Designated initializers are strongly recommended here. You can drop all the NULLs as the language guarantees uninitialized members are 0.

File: ompi/mca/part/base/part_base_precvreq.c [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#include "ompi/types.h"
+#include "ompi/mca/part/part.h"
+#include "ompi/mca/part/base/part_base_precvreq.h"
+
+
+static void mca_part_base_precv_request_construct(mca_part_base_precv_request_t*);
+static void mca_part_base_precv_request_destruct(mca_part_base_precv_request_t*);
+
+
+OBJ_CLASS_INSTANCE(
```
Comment (hjelmn) [inline ompi/mca/part/base/part_base_precvreq.c:31]: nit: you could move this after the constructor/destructor definitions and avoid the declarations above.

File: ompi/mca/part/base/part_base_precvreq.c [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+);
+
+
+static void mca_part_base_precv_request_construct(mca_part_base_precv_request_t* request)
+{
+    OBJ_CONSTRUCT(&request->req_base.req_convertor, opal_convertor_t);
+}
+
+
+static void mca_part_base_precv_request_destruct(mca_part_base_precv_request_t* request)
```
Comment (hjelmn) [inline ompi/mca/part/base/part_base_precvreq.c:45]: nit: You can move this comment above the OBJ_CLASS_INSTANCE and remove the destructor.

File: ompi/mca/part/base/part_base_psendreq.c [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+#include "ompi_config.h"
+#include <string.h>
+#include "ompi/mca/part/part.h"
+#include "ompi/mca/part/base/part_base_psendreq.h"
+
+static void mca_part_base_psend_request_construct(mca_part_base_psend_request_t* req);
+static void mca_part_base_psend_request_destruct(mca_part_base_psend_request_t* req);
+
+
+OBJ_CLASS_INSTANCE(
```
Comment (hjelmn) [inline ompi/mca/part/base/part_base_psendreq.c:28]: Same comment as above. You can clean this up a little.

File: ompi/mca/part/persist/part_persist.c [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "ompi/communicator/communicator.h"
+#include "ompi/mca/part/base/part_base_prequest.h"
+#include "ompi/mca/part/base/base.h"
+
+#include "ompi/mca/part/persist/part_persist.h"
+#include "ompi/mca/part/persist/part_persist_sendreq.h"
+#include "ompi/mca/part/persist/part_persist_recvreq.h"
+
+ompi_part_persist_t ompi_part_persist = {
```
Comment (hjelmn) [inline ompi/mca/part/persist/part_persist.c:31]: Please use designated initializers here.

File: ompi/mca/part/persist/part_persist.h [hunk 6]
```diff
@@ -0,0 +1,595 @@
...
+    if(-1 == ompi_part_persist.init_world)
+    {
+        OPAL_THREAD_UNLOCK(&ompi_part_persist.lock);
+        block_entry = opal_atomic_add_fetch_32(&(ompi_part_persist.block_entry), -1);
+        return OMPI_SUCCESS;
+    }
+
+    /* Can't do anything if we don't have world */
+    if(0 == ompi_part_persist.init_world) {
+        int err = MPI_Comm_rank(MPI_COMM_WORLD, &ompi_part_persist.my_world_rank);
```
Comment (hjelmn) [inline ompi/mca/part/persist/part_persist.h:185]: Should change these to use the internal versions of these functions. ompi_comm vs MPI_Comm

File: ompi/mca/part/persist/part_persist_sendreq.h [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+#define MCA_PART_PERSIST_PSEND_REQUEST_ALLOC(sendreq, comm, dst,          \
+                                           ompi_proc)                 \
+do {                                                                  \
+    sendreq = (mca_part_persist_psend_request_t*)                         \
+        opal_free_list_wait (&mca_part_base_psend_requests);          \
+    sendreq->req_base.req_type = MCA_PART_PERSIST_REQUEST_PSEND;          \
+} while(0)
+
+#define MCA_PART_PERSIST_PSEND_REQUEST_INIT( req_send,                    \
```
Comment (hjelmn) [inline ompi/mca/part/persist/part_persist_sendreq.h:49]: nit: At this point I believe we recommend always-inline static inline vs macros. I am fine with these they way they are but it does somewhat hurt readability.

File: ompi/mpi/c/pready_list.c [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        rc = OMPI_SUCCESS;
+
+        OMPI_ERR_INIT_FINALIZE(FUNC_NAME);
+        if (NULL == request || OMPI_REQUEST_PART != request->req_type) {
+            rc = MPI_ERR_REQUEST;
+        }
+        OMPI_ERRHANDLER_CHECK(rc, MPI_COMM_WORLD, rc, FUNC_NAME);
+    }
+
+    for(i = 0; i < length && OMPI_SUCCESS == rc; i++) {
```
Comment (hjelmn) [inline ompi/mpi/c/pready_list.c:62]: i is not used outside this look. move the declaration into the for statement.

File: ompi/mpi/fortran/mpif-h/pready_list_f.c [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+#endif
+#endif
+
+
+void ompi_pready_list_f(MPI_Fint *length, MPI_Fint *partitions, MPI_Fint *request, MPI_Fint *ierr)
+{
+   int c_ierr, i;
+   MPI_Request c_req = PMPI_Request_f2c(*request);
+   int *c_partitions = (int*)malloc(sizeof(int)*OMPI_FINT_2_INT(*length));
+   for(i = 0; i < OMPI_FINT_2_INT(*length); i++)
```
Comment (hjelmn) [inline ompi/mpi/fortran/mpif-h/pready_list_f.c:76]: Move declaration of i into the loop initalizer.

**Review by hjelmn — CHANGES_REQUESTED** [review]

**Comment by mdosanjh** [comment]
bot:retest

File: ompi/mca/part/persist/part_persist_sendreq.h (hunk 7 above)
Comment (mdosanjh) [reply ompi/mca/part/persist/part_persist_sendreq.h:49]: @hjelmn The most recent commit addresses all of the requested changes except for this one. If this is a blocker I can  go back and change the macros, but my inclination is to leave them as is for now if that works.

**Comment by mdosanjh** [comment]
bot:retest

**Comment by mdosanjh** [comment]
bot:retest

**Comment by mdosanjh** [comment]
bot:retest

**Comment by jsquyres** [comment]
@mdosanjh Please see https://github.com/open-mpi/ompi/wiki/PRJenkins for a listing of the bot commands.  If possible, please be judicious with re-running CI, because our CI can get backed up with queued builds, etc.

**Comment by jsquyres** [comment]
Oops -- clicked the wrong button there; didn't mean to close this PR.

**Comment by mdosanjh** [comment]
> @mdosanjh Please see https://github.com/open-mpi/ompi/wiki/PRJenkins for a listing of the bot commands.  If possible, please be judicious with re-running CI, because our CI can get backed up with queued builds, etc.

Apologies, the build kept failing on ARMV8 with a GitHub check out error (with a reference a commit (other than mine) not being a tree). I'll try to be more specific with the bot commands in the future.

File: ompi/include/mpi.h.in [hunk 10]
```diff
@@ -2310,6 +2321,17 @@ OMPI_DECLSPEC  int PMPI_Isend(const void *buf, int count, MPI_Datatype datatype,
                               int tag, MPI_Comm comm, MPI_Request *request);
 OMPI_DECLSPEC  int PMPI_Issend(const void *buf, int count, MPI_Datatype datatype, int dest,
                                int tag, MPI_Comm comm, MPI_Request *request);
+OMPI_DECLSPEC  int PMPI_Precv_init(const void* buf, int partitions, MPI_Count count,
```
Comment (hppritcha) [inline ompi/include/mpi.h.in:2324]: same here

File: ompi/include/mpi.h.in [hunk 11]
```diff
@@ -1645,6 +1645,17 @@ OMPI_DECLSPEC  int MPI_Isend(const void *buf, int count, MPI_Datatype datatype,
                              int tag, MPI_Comm comm, MPI_Request *request);
 OMPI_DECLSPEC  int MPI_Issend(const void *buf, int count, MPI_Datatype datatype, int dest,
                               int tag, MPI_Comm comm, MPI_Request *request);
+OMPI_DECLSPEC  int MPI_Precv_init(const void* buf, int partitions, MPI_Count count,
```
Comment (hppritcha) [inline ompi/include/mpi.h.in:1648]: could you please add these in alphabetical order?

**Review by hppritcha — CHANGES_REQUESTED** [review]

File: ompi/request/request_dbg.h [hunk 12]
```diff
@@ -25,6 +25,7 @@ typedef enum {
     OMPI_REQUEST_GEN,      /**< MPI-2 generalized request */
     OMPI_REQUEST_WIN,      /**< MPI-2 one-sided request */
     OMPI_REQUEST_COLL,     /**< MPI-3 non-blocking collectives request */
+    OMPI_REQUEST_PART,     /**< MPI-4 partitioned communication request */
```
[the remark this answers is not in the document]
Comment (mdosanjh) [reply ompi/request/request_dbg.h]: @hppritcha Thank you for the review, I've fixed all of the alphabetical orderings and as per this comment, there was not a reason for the placement, so I have moved OMPI_REQUEST_PART to be after OMPI_REQUEST_COMM.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/757b00059d0d4eb755171f50e177e95e

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/7f9273003d2846772017c4643f28d9c4

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/945f245ddfc09f44492840d286a255ba

**Comment by hppritcha** [comment]
@mdosanjh could you rebase this PR?  Looks like a simple conflict.

**Comment by mdosanjh** [comment]
> @mdosanjh could you rebase this PR?  Looks like a simple conflict.

@hppritcha It should be rebased now.

**Review by hppritcha — APPROVED** [review]

**Comment by mdosanjh** [comment]
@hjelmn I just wanted check to see if you've gotten a chance to re-review this since I incorporated your requested changes.

**Comment by mdosanjh** [comment]
> @hjelmn I just wanted check to see if you've gotten a chance to re-review this since I incorporated your requested changes. 

It was suggested that I put some sort of timeline here to say when I would 'dismiss' the review, specifically the 23rd. (If this is unreasonable, please let me know)

**Review by hjelmn — APPROVED** [review]

**Comment by hppritcha** [comment]
@regrant does this PR include all partitioned comm related items that went in to MPI 4.0?

**Comment by regrant** [comment]
@hppritcha This has all of the MPI 4.0 related content, so we're 4.0 compliant for partitioned communication with this merge.

**Pull request merged**